File: lib/models/group.dart
import 'user.dart';

enum GroupMemberRole {
  admin,
  member,
}

class GroupMember {
  final User user;
  final GroupMemberRole role;
  final String? nickname; // Custom nickname in this group
  final bool isVisible; // If false, member is "invisible" (privacy feature)
  final DateTime joinedAt;

  GroupMember({
    required this.user,
    required this.role,
    this.nickname,
    this.isVisible = true,
    required this.joinedAt,
  });

  factory GroupMember.fromJson(Map<String, dynamic> json) {
    return GroupMember(
      user: User.fromJson(json['user']),
      role: GroupMemberRole.values.firstWhere(
        (e) => e.toString() == 'GroupMemberRole.${json['role']}',
        orElse: () => GroupMemberRole.member,
      ),
      nickname: json['nickname'],
      isVisible: json['isVisible'] ?? true,
      joinedAt: DateTime.parse(json['joinedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user': user.toJson(),
      'role': role.toString().split('.').last,
      'nickname': nickname,
      'isVisible': isVisible,
      'joinedAt': joinedAt.toIso8601String(),
    };
  }
}

class Group {
  final String id;
  final String name;
  final String? description;
  final String? image;
  final List<GroupMember> members;
  final bool isPrivate;
  final bool allowAnonymousMessages;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int? messageExpiryTime; // Time in seconds for messages to expire

  Group({
    required this.id,
    required this.name,
    this.description,
    this.image,
    required this.members,
    required this.isPrivate,
    this.allowAnonymousMessages = false,
    required this.createdAt,
    required this.updatedAt,
    this.messageExpiryTime,
  });

  factory Group.fromJson(Map<String, dynamic> json) {
    return Group(
      id: json['_id'] ?? json['id'],
      name: json['name'],
      description: json['description'],
      image: json['image'],
      members: (json['members'] as List)
          .map((member) => GroupMember.fromJson(member))
          .toList(),
      isPrivate: json['isPrivate'] ?? true,
      allowAnonymousMessages: json['allowAnonymousMessages'] ?? false,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      messageExpiryTime: json['messageExpiryTime'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'image': image,
      'members': members.map((member) => member.toJson()).toList(),
      'isPrivate': isPrivate,
      'allowAnonymousMessages': allowAnonymousMessages,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'messageExpiryTime': messageExpiryTime,
    };
  }

  List<GroupMember> get admins {
    return members.where((member) => member.role == GroupMemberRole.admin).toList();
  }

  List<GroupMember> get visibleMembers {
    return members.where((member) => member.isVisible).toList();
  }

  bool isAdmin(String userId) {
    final member = members.firstWhere(
      (member) => member.user.id == userId,
      orElse: () => GroupMember(
        user: User(
          id: '',
          phoneNumber: '',
          createdAt: DateTime.now(),
          lastSeen: DateTime.now(),
          publicKey: {},
        ),
        role: GroupMemberRole.member,
        joinedAt: DateTime.now(),
      ),
    );
    return member.role == GroupMemberRole.admin;
  }

  String? getMemberNickname(String userId) {
    final member = members.firstWhere(
      (member) => member.user.id == userId,
      orElse: () => GroupMember(
        user: User(
          id: '',
          phoneNumber: '',
          createdAt: DateTime.now(),
          lastSeen: DateTime.now(),
          publicKey: {},
        ),
        role: GroupMemberRole.member,
        joinedAt: DateTime.now(),
      ),
    );
    return member.nickname;
  }
}

File: server/models/group.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;

const groupMemberSchema = new Schema({
  user: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  role: {
    type: String,
    enum: ['admin', 'member'],
    default: 'member',
  },
  nickname: {
    type: String,
    trim: true,
  },
  isVisible: {
    type: Boolean,
    default: true,
  },
  joinedAt: {
    type: Date,
    default: Date.now,
  },
});

const groupSchema = new Schema({
  name: {
    type: String,
    required: true,
    trim: true,
  },
  description: {
    type: String,
    trim: true,
  },
  image: {
    type: String,
  },
  members: [groupMemberSchema],
  isPrivate: {
    type: Boolean,
    default: true,
  },
  allowAnonymousMessages: {
    type: Boolean,
    default: false,
  },
  messageExpiryTime: {
    type: Number, // Time in seconds for messages to expire
  },
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  chatId: {
    type: Schema.Types.ObjectId,
    ref: 'Chat',
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
  updatedAt: {
    type: Date,
    default: Date.now,
  },
}, {
  timestamps: true
});

// Virtual for getting ID as a string
groupSchema.virtual('id').get(function() {
  return this._id.toHexString();
});

// Ensure virtual fields are serialized
groupSchema.set('toJSON', {
  virtuals: true,
  transform: (doc, ret) => {
    delete ret._id;
    delete ret.__v;
    return ret;
  }
});

// Method to check if a user is a member of the group
groupSchema.methods.isMember = function(userId) {
  return this.members.some(
    member => member.user.toString() === userId.toString()
  );
};

// Method to check if a user is an admin of the group
groupSchema.methods.isAdmin = function(userId) {
  return this.members.some(
    member => member.user.toString() === userId.toString() && member.role === 'admin'
  );
};

// Method to add a member to the group
groupSchema.methods.addMember = function(userId, options = {}) {
  if (this.isMember(userId)) {
    return Promise.resolve(this);
  }
  
  this.members.push({
    user: userId,
    role: options.isAdmin ? 'admin' : 'member',
    nickname: options.nickname,
    isVisible: options.isVisible !== undefined ? options.isVisible : true,
    joinedAt: new Date(),
  });
  
  this.updatedAt = new Date();
  return this.save();
};

// Method to remove a member from the group
groupSchema.methods.removeMember = function(userId) {
  this.members = this.members.filter(
    member => member.user.toString() !== userId.toString()
  );
  
  this.updatedAt = new Date();
  return this.save();
};

// Method to update a member's settings
groupSchema.methods.updateMember = function(userId, updates) {
  const member = this.members.find(
    member => member.user.toString() === userId.toString()
  );
  
  if (!member) {
    return Promise.resolve(this);
  }
  
  if (updates.nickname !== undefined) {
    member.nickname = updates.nickname;
  }
  
  if (updates.isVisible !== undefined) {
    member.isVisible = updates.isVisible;
  }
  
  if (updates.role !== undefined) {
    member.role = updates.role;
  }
  
  this.updatedAt = new Date();
  return this.save();
};

// Method to get a member's details
groupSchema.methods.getMember = function(userId) {
  return this.members.find(
    member => member.user.toString() === userId.toString()
  );
};

module.exports = mongoose.model('Group', groupSchema);